### PR TITLE
Add PV latest forecast in uk map

### LINKED ISSUE #23: Add latest API data to Map
# Add the latest forecast to UK Map

Screenshot 2022-05-09 at 09 55 05

- [x] The map should be zoomable

- hover over region should give PV value (needs designs)

API end point:
- v0/GB/solar/gsp/forecast/all
Note currently this takes a few seconds to load
This provider's 338 forecast (ignore the gsp_id=0)

```{
  "forecasts": [
    {
      "location": {
        "label": "string",
        "gspId": 0,
        "gspName": "string",
        "gspGroup": "string",
        "regionName": "string",
        "rmMode": true
      },
      "model": {
        "name": "string",
        "version": "string"
      },
      "forecastCreationTime": "2022-05-09T08:56:38.870Z",
      "forecastValues": [
        {
          "targetTime": "2022-05-09T08:56:38.871Z",
          "expectedPowerGenerationMegawatts": 0
        }
      ],
      "inputDataLastUpdated": {
        "gsp": "2022-05-09T08:56:38.871Z",
        "nwp": "2022-05-09T08:56:38.871Z",
        "pv": "2022-05-09T08:56:38.871Z",
        "satellite": "2022-05-09T08:56:38.871Z"
      }
    }
  ]
}
```
We should use the first `forecastValues` and use `expectedPowerGenerationMegawatts` for the value

## Colours: 
- yellow (hex in experiments folder)
- opacity, start at 0, going to 100%
- range 0 to 500 kw (would be good to be able to easily change the maximum, in the code)

## Refresh
- data should be refreshed every 5 mins. In the future, we can adjust the time the fresh is made

Make it vertical (northernly aligned), but have an easy way to change to a bearing of 90 (in the code)

## Future: 
Want to show animation / video of different forecasts, i.e 30 mins, 1 hour, 1.5 hours and 2 hours.

### PR DESCRIPTION
**Pull request by Alaatohamy**
closes #23
## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/nowcasting/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: apps/nowcasting-app/pages/index.tsx [hunk 1]
```diff
@@ -5,52 +5,102 @@ import useSWR from "swr";
...
 import ButtonGroup from "../components/button-group";
+import gspShapeData from "../data/gsp-regions.json";
+import * as d3 from "d3";
+
 
 const fetcher = (input: RequestInfo, init: RequestInit) =>
   fetch(input, init).then((res) => res.json());
 
-const API_PREFIX = "https://api-dev.nowcasting.io/v0";
+const API_PREFIX = "http://nowcasting-api-development.eu-west-1.elasticbeanstalk.com/v0";
```
Comment (dlinah) [inline apps/nowcasting-app/pages/index.tsx:15]: move to env var or constant file

File: apps/nowcasting-app/pages/index.tsx [hunk 2]
```diff
@@ -5,52 +5,102 @@ import useSWR from "swr";
...
+const latestForecastValue = 0;
 
 export default function Home() {
+
   const { data: forecastData, error: forecastError } = useSWR(
     `${API_PREFIX}/GB/solar/gsp/forecast/all`,
-    fetcher
+    fetcher,
+    // every 5 minutes
+    { refreshInterval: 300000 }
```
Comment (dlinah) [inline apps/nowcasting-app/pages/index.tsx:26]: call update map

**Review by dlinah — COMMENTED** [review]
LGTM 💯

File: apps/nowcasting-app/pages/index.tsx [hunk 3]
```diff
@@ -4,53 +4,121 @@ import useSWR from "swr";
...
+    })
+   
+    const extent = d3.extent([...allValues]);
+
+    return {     
+      "fill-color": "#eab308",
+      "fill-opacity": 
+      [
+        "interpolate",
+        ["exponential", 1],
```
Comment (dctanner) [inline apps/nowcasting-app/pages/index.tsx:63]: This `fill-opacity` array defines how the yellow gsp opacity levels are chosen, based on the expectedPowerGenerationMegawatts values (see https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#interpolate).

From what I understand, this code uses the highest expectedPowerGenerationMegawatts as the highest value in the opacity scale (`extent[1]`). However in the story @peterdudfield proposed we use a fixed value of 400MW as the highest value. @Alaatohamy I believe we will need to change `extent[1]` to be `400` (probably saved in constant.ts).

Comment (dctanner) [inline apps/nowcasting-app/pages/index.tsx:63]: We currently use an exponential for the opacity scale (see https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#interpolate). @peterdudfield do you think linear or exponential makes more sense?

File: apps/nowcasting-app/pages/index.tsx [hunk 4]
```diff
@@ -4,53 +4,121 @@ import useSWR from "swr";
...
+        ['get', 'expectedPowerGenerationMegawatts'],
+        extent[0],
+        0,
+        extent[1],
```
Comment (dctanner) [inline apps/nowcasting-app/pages/index.tsx:67]: This `fill-opacity` array defines how the yellow gsp opacity levels are chosen, based on the expectedPowerGenerationMegawatts values (see https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#interpolate).

From what I understand, this code uses the highest expectedPowerGenerationMegawatts as the highest value in the opacity scale (`extent[1]`). However in the story @peterdudfield proposed we use a fixed value of 400MW as the highest value. @Alaatohamy I believe we will need to change `extent[1]` to be `400` (probably saved in constant.ts).

**Comment by dctanner** [comment]
@Alaatohamy It might be cleaner if we moved out most of the map and data fetching related code from `index.tsx` to a file in a new folder like `components/map/`

**Comment by dctanner** [comment]
Changes from call today:

- [ ] Change fill-opacity scale to linear, with 500 maximum value
- [ ] Show blank map when loading. No need for loading indicator anywhere. Simply update the map with new data when it's loaded in the background.

We're ready to merge to prod once we've done these and the other comments.

**Comment by dctanner** [comment]
@Alaatohamy please also add the standard mapbox zoom co
CleanShot 2022-05-16 at 11 15 44@2x
ntrols to the map too like this

File: apps/nowcasting-app/components/map/map.tsx [hunk 5]
```diff
@@ -45,9 +47,16 @@ const Map = ({ loadDataOverlay, controlOverlay, bearing = 90 }: IMap) => {
...
       loadDataOverlay(map);
     });
+
+    // Indecated that data source updates every 5 minutes
+    // map.current.on("sourcedata", (event) => {
+    //   if(event.isSourceLoaded && event.sourceCacheId && event.source.data){
+    //     console.log('A data event occurred.', event);
+    //   }
+    // });
   });
```
Comment (dlinah) [inline apps/nowcasting-app/components/map/map.tsx:60]: @Alaatohamy should be dependent  on map (maybe that is why it was loading source twice)
```suggestion
  },[map]);
```

File: apps/nowcasting-app/components/map/map.tsx [hunk 6]
```diff
@@ -45,10 +47,17 @@ const Map = ({ loadDataOverlay, controlOverlay, bearing = 90 }: IMap) => {
...
       loadDataOverlay(map);
     });
-  });
+
+    // Indecated that data source updates every 5 minutes
+    // map.current.on("sourcedata", (event) => {
+    //   if(event.isSourceLoaded && event.sourceCacheId && event.source.data){
+    //     console.log('A data event occurred.', event);
+    //   }
+    // });
```
Comment (dlinah) [inline apps/nowcasting-app/components/map/map.tsx:59]:
```suggestion

```

**Review by dlinah — APPROVED** [review]
💯

**Pull request merged**